src/wp-includes/class-wp-theme-json.php

```diff
@@ -729,14 +729,19 @@ protected static function append_to_selector( $selector, $to_append, $position =
 	 * @return array Block metadata.
 	 */
 	protected static function get_blocks_metadata() {
-		if ( null !== static::$blocks_metadata ) {
-			return static::$blocks_metadata;
-		}
-
-		static::$blocks_metadata = array();
-
 		$registry = WP_Block_Type_Registry::get_instance();
 		$blocks   = $registry->get_all_registered();
+
+		if ( null === static::$blocks_metadata ) {
+			static::$blocks_metadata = array();
+		} else {
+			// Do we have metadata for all currently registered blocks?
+			$blocks = array_diff_key( $blocks, static::$blocks_metadata );
+			if ( count( $blocks ) === 0 ) {
+				return static::$blocks_metadata;
+			}
+		}
+
```

Comment:
Adding an explanation I posted over at https://github.com/WordPress/gutenberg/pull/44658#issuecomment-1266574880 to clarify how this works:

> This mostly leverages PHP's [`array_diff_key`](https://www.php.net/manual/en/function.array-diff-key.php): It looks at the list of currently registered `$blocks` and compares its keys (which are block names, e.g. `core/heading`) to those of the list of blocks metadata -- which are also keyed by block name. It then returns the subset of registered `$blocks` whose keys aren't in `static::$blocks_metadata`. If that subset is empty, we return the cached blocks metadata, as we can assume that we have metadata for all registered blocks already.
> 
> Otherwise, we add metadata -- for that subset of registered blocks only that doesn't have them yet.

Reply:
Importantly, this means that cache invalidation is _not_ based on the number of blocks, but indeed on _which_ blocks have been registered.

Reply:
In https://github.com/WordPress/wordpress-develop/pull/3373/#discussion_r986628465, @ajlende noted that

> Unfortunately we can't rely on only the registered block names because there is at least one instance where we unregister an old version of a block and re-register a new one with the same name within the same function.
>
> https://github.com/WordPress/wordpress-develop/pull/3373#issuecomment-1263888253

_tl;dr:_ I _think_ that in practice, the list of registered block names is a good enough indicator for caching.

The example that's present in Core, [`register_legacy_post_comments_block`](https://github.com/WordPress/wordpress-develop/blob/4a3d6ff8213d5264ba36c4021f8b9f0be227da51/src/wp-includes/blocks/comments.php#L152), is basically used to replace occurrences of the legacy Post Comments block with its successor (the Comments block), in "legacy" (compatibility) mode. 

However, while this logic was relevant to have in the Gutenberg plugin while Core still included the legacy Post Comments block, the latter is no longer the case -- it has been [scrapped from Core](https://github.com/WordPress/wordpress-develop/pull/3154/files#diff-1006e616f17f4792802a55871dd42b7057c232377ec8fbc750a60ad850cf702a) per https://github.com/WordPress/wordpress-develop/pull/3154. So the [unregistering logic](https://github.com/WordPress/wordpress-develop/blob/4a3d6ff8213d5264ba36c4021f8b9f0be227da51/src/wp-includes/blocks/comments.php#L155-L161) will never kick in.

And even if it did, the metadata we're using to (re-)register it are the same that the Post Comments block was using.

This brings us to the bigger question: Can this situation -- a block is registered, then unregistered, and then re-registered under the same name _but with different metadata_ -- happen in principle? I think the chance is fairly low. While we've seen one precedent of block re-registration in Core (where the functionality of one block was absorbed into another, but the original block hat to be retained for back-compat), this is something that isn't going to happen very often; and if it is, we're probably going to use the same metadata to register it.

I think the same would apply to plugins. I don't think there are a lot of plugins that will register a block only to un-register and then re-register it; what's more conceivable is that a _different_ plugin does the un-registration and re-registration bit. (Generally, I'd expect that other plugin to be from the same author -- e.g. something like [WooCommerce Blocks](https://de.wordpress.org/plugins/woo-gutenberg-products-block/) which can be used "on top of" WooCommerce. Otherwise, it's debatable if a plugin should really touch another plugin's block namespace that way at all.)

Re-registering blocks from a different author or even Core seems so risky that I don't think we need to provision for that case: That existing block's attributes (and other metadata) basically define the block's "interface", so any replacement would have to keep that intact. This doesn't seem to leave a lot of conceivable use cases.